test/extended/networking/ovs.go

```diff
@@ -0,0 +1,314 @@
+					if strings.Contains(flow, "nw_dst="+ip) {
+						foundServiceFlow = true
+						break
+					}
+				}
+				Expect(foundServiceFlow).To(BeTrue(), "Each node contains a rule for the service")
+			}
+
+			err := f1.Client.Pods(f1.Namespace.Name).Delete(serviceName, nil)
+			Expect(err).NotTo(HaveOccurred())
+			err = f1.Client.Services(f1.Namespace.Name).Delete(serviceName)
+			Expect(err).NotTo(HaveOccurred())
+
+			postDeleteFlows := getFlowsForAllNodes(oc, nodes.Items)
+			Expect(reflect.DeepEqual(origFlows, postDeleteFlows)).To(BeTrue(), "Flows after deleting service should be same as before creating it")
+		})
+	})
+})
+
+func doGetFlowsForNode(oc *testexutil.CLI, nodeName string) ([]string, error) {
+	pod := &kapi.Pod{
+		TypeMeta: kapiunversioned.TypeMeta{
+			Kind: "Pod",
+		},
+		ObjectMeta: kapi.ObjectMeta{
+			GenerateName: "flow-check",
+		},
+		Spec: kapi.PodSpec{
+			Containers: []kapi.Container{
+				{
+					Name:  "flow-check",
+					Image: "openshift/openvswitch",
+					// kubernetes seems to get confused sometimes if the pod exits too quickly
+					Command: []string{"sh", "-c", "ovs-ofctl -O OpenFlow13 dump-flows br0 && sleep 1"},
+					VolumeMounts: []kapi.VolumeMount{
+						{
+							Name:      "ovs-socket",
+							MountPath: "/var/run/openvswitch/br0.mgmt",
+						},
+					},
+				},
+			},
+			Volumes: []kapi.Volume{
+				{
+					Name: "ovs-socket",
+					VolumeSource: kapi.VolumeSource{
+						HostPath: &kapi.HostPathVolumeSource{
+							Path: "/var/run/openvswitch/br0.mgmt",
+						},
+					},
+				},
+			},
+			NodeName:      nodeName,
+			RestartPolicy: kapi.RestartPolicyNever,
+			// We don't actually need HostNetwork, we just set it so that deploying this pod won't cause any OVS flows to be added
+			SecurityContext: &kapi.PodSecurityContext{
+				HostNetwork: true,
+			},
+		},
+	}
+	f := oc.KubeFramework()
+	podClient := f.Client.Pods(f.Namespace.Name)
+	pod, err := podClient.Create(pod)
+	if err != nil {
+		return nil, err
+	}
+	defer podClient.Delete(pod.Name, nil)
+	err = waitForPodSuccessInNamespace(f.Client, pod.Name, "flow-check", f.Namespace.Name)
+	if err != nil {
+		return nil, err
+	}
+	logs, err := oc.Run("logs").Args(pod.Name).Output()
+	if err != nil {
+		return nil, err
+	}
+
+	// For ease of comparison, strip out the parts of the rules that change
+	flows := strings.Split(logs, "\n")
+	strip_re := regexp.MustCompile(`(duration|n_packets|n_bytes)=[^,]*, `)
+	for i := range flows {
+		flows[i] = strip_re.ReplaceAllLiteralString(flows[i], "")
+	}
+	return flows, nil
+}
+
+func getFlowsForNode(oc *testexutil.CLI, nodeName string) []string {
+	flows, err := doGetFlowsForNode(oc, nodeName)
+	expectNoError(err)
+	return flows
+}
+
+func getFlowsForAllNodes(oc *testexutil.CLI, nodes []kapi.Node) map[string][]string {
+	var err error
+	flows := make(map[string][]string, len(nodes))
+	for _, node := range nodes {
+		flows[node.Name], err = doGetFlowsForNode(oc, node.Name)
```

Comment:
expectNoError(err) is a function in util.go that uses ginkgo/gomega magic so that if the expectation fails, it gets reported as a failure in the function's caller rather than the function itself. So the reason this is split up the way it is is so that both getFlowsForNode() failures and getFlowsForAllNodes() failures both get reported with the appropriate line number from the main test rather than the line numbers from the helper functions